File: October2022/Java/DeleteTheMiddleNodeOfALinkedList.java

// @saorav21994

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        
        if (head == null || head.next == null)
            return null;
        
        if (head.next.next == null) {   // only 2 node exist
            head.next = null;
            return head;
        }
        
        ListNode slow = head;
        ListNode fast = head.next.next;
        
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
            if (fast.next != null)
                fast = fast.next;
        }
        
        slow.next = slow.next.next;
        
        return head;
            
    }
}

File: October2022/Python/DeleteTheMiddleNodeOfALinkedList.py
# https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list
# Author : @Jiganesh 

from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:

        dummyNode = ListNode(-1)
        dummyNode.next = head
        
        prev = dummyNode
        slow = fast = head

        while fast and fast.next:
            prev, slow, fast = slow, slow.next , fast.next.next

        prev.next = slow.next
        return dummyNode.next
